### PR TITLE
Add upper limit on numberOfPage and PagesRead

### PR DESCRIPTION
**Pull request by prabhjot018**
## Add upper limit on numberOfPage and PagesRead in book

Add max constraint on numberOfPage and PagesRead members in Book entity

## Additional context

Not a UI change

## Related issue

https://github.com/knjk04/book-project/issues/258

## Pull request checklist

Please ensure you have done the following:

- [x] Read, understood and adhered to our [contributing document](https://github.com/knjk04/book-project/blob/master/CONTRIBUTING.md).
  - [x] Ensure that you were first assigned to a relevant issue before creating this pull request
  - [x] Ensure code changes pass all tests

- [x] Read, understood and adhered to our [style guide](https://github.com/knjk04/book-project/blob/master/STYLEGUIDE.md). A lot of our code reviews are spent on ensuring compliance with our style guide, so it would save a lot of time if this was adhered to from the outset. 

- [x] Filled in the summary, context (if applicable) and related issue section. Replace the square brackets and its placeholder content with your contents. For an example, see any merged in pull request
  - [x] Included a screenshot(s) if a UI change was involved (it may look different on a reviewer's device)

- [x] Created a branch that has a descriptive name (what your branch is for a in a few words and includes the issue number at the end, e.g. `test-reading-goal-123`

- [x] Set this pull request to 'draft' if you are still working on it

- [x] Resolved any merge conflicts

If in doubt, get in touch with us via our [Gitter room](https://gitter.im/book-project-community/community)

### CODE REVIEW

**Comment by knjk04** [comment]
When you're ready for a review, let me know by clicking on the 'ready to review' button. Before you do that, please ensure you fill out the checklist

File: src/main/java/com/karankumar/bookproject/backend/entity/Book.java
Comment (knjk04) [inline src/main/java/com/karankumar/bookproject/backend/entity/Book.java:48]: Instead of using the same value twice, could you create a constant (called something like 'max')? Then, we can use this in the test as well if you set the scope to the default package-level access and mark it as `@VisibleForTesting`.

Also, when you create this max variable, could you use an underscore to separate the thousandths (see the [numeric literals section in our styleguide](https://github.com/knjk04/book-project/blob/master/STYLEGUIDE.md#underscores-in-numeric-literals))

File: src/test/java/com/karankumar/bookproject/backend/service/BookServiceTest.java
Comment (knjk04) [inline src/test/java/com/karankumar/bookproject/backend/service/BookServiceTest.java:92]: Instead of testing `Integer.MAX_VALUE`, could you test the one number above the boundary (23,001)?

Also, could you add tests for the boundary and one number less than the boundary?

File: src/test/java/com/karankumar/bookproject/backend/service/BookServiceTest.java
Comment (knjk04) [inline src/test/java/com/karankumar/bookproject/backend/service/BookServiceTest.java:69]: Could you add a method (either a constructor. a `@BeforeEach` or after each setup method) that removes all of the books?

**Review by knjk04 — CHANGES_REQUESTED** [review]
Thanks for your implementation and creating a new PR. It's nice that you used the `@Max` constraint annotation and I like that you're getting the root cause of the exception.

In addition to addressing the review comments, could you also a validation check on the frontend to ensure that in the event that user enters an invalid number for the number of pages or pages read, there is a form error and they are not allowed to proceed? If you need a hand with this, let me know and I can point you in the right direction

**Comment by knjk04** [comment]
Also, I tried to set a book that had 22,999 pages, but I was getting a null pointer exception. Do you also have this problem?

File: src/test/java/com/karankumar/bookproject/backend/service/BookServiceTest.java
Comment (amit1307) [inline src/test/java/com/karankumar/bookproject/backend/service/BookServiceTest.java:84]: Instead of the catch block do you want to use something like below:

`Throwable exception = Assertions.catchThrowable(() -> bookService.save(book))`

Then you can assert the exception variable for the exception type, would be better if we can assert that error message as well if there is any.

File: src/test/java/com/karankumar/bookproject/backend/service/BookServiceTest.java
Comment (amit1307) [inline src/test/java/com/karankumar/bookproject/backend/service/BookServiceTest.java:101]: Same as the above comment, would be better if we can get rid of the try catch.

**Review by amit1307 — CHANGES_REQUESTED** [review]

**Comment by knjk04** [comment]
Thanks for the review @amit1307 

@prabhjot018 Could you please fix the merge conflict and then us know when you need another review?

File: src/main/java/com/karankumar/bookproject/backend/entity/Book.java
Comment (prabhjot018) [reply src/main/java/com/karankumar/bookproject/backend/entity/Book.java:48]: Surprisingly I can have only constant expression in the annotation, I tried creating a Constant class like

`public final class Constants {
    public static final Integer MAX_PAGES = 23_000;
}
`
and used this in max annotation like 
` @Max(value = Constants.MAX_PAGES)
    private Integer numberOfPages;`

But I am getting java error - java element value must be a constant expression

Comment (prabhjot018) [reply src/main/java/com/karankumar/bookproject/backend/entity/Book.java:48]: @knjk04 wdyt?

**Comment by prabhjot018** [comment]
> Also, I tried to set a book that had 22,999 pages, but I was getting a null pointer exception. Do you also have this problem?

No, I didn't face this issue

**Comment by prabhjot018** [comment]
@knjk04 @amit1307 Hey! I have addressed review comments, please have a look.

File: src/main/java/com/karankumar/bookproject/backend/entity/Book.java
Comment (amit1307) [reply src/main/java/com/karankumar/bookproject/backend/entity/Book.java:52]: If I understood Karan's original comment, he might mean just create a constant like below:

`public static final Integer MAX_PAGES = 23_000;`

in the same class and annotate it with `@VisibleForTesting`

File: src/test/java/com/karankumar/bookproject/backend/service/BookServiceTest.java
Comment (amit1307) [inline src/test/java/com/karankumar/bookproject/backend/service/BookServiceTest.java:101]: Its better but would be better if we can try to make it bit more simple what I had suggested in my original comment like:

`Throwable error = Assertions.catchThrowable(() -> bookService.save(book))`

**Comment by amit1307** [comment]
> @knjk04 @amit1307 Hey! I have addressed review comments, please have a look.

Hey just a quick comment, could you please leave the comments unresolved and leave that to the author of the comment?

File: src/test/java/com/karankumar/bookproject/backend/service/BookServiceTest.java
Comment (prabhjot018) [reply src/test/java/com/karankumar/bookproject/backend/service/BookServiceTest.java:101]: The issue was Assertions is already imported from the different package and I have to use a fully qualified package name for it to be used like - 
`org.assertj.core.api.Assertions.catchThrowable()`

That's why I wrote the code like that, if you feel this is okay I will change

File: src/main/java/com/karankumar/bookproject/backend/entity/Book.java
Comment (prabhjot018) [reply src/main/java/com/karankumar/bookproject/backend/entity/Book.java:52]: `public class Book extends BaseEntity {
    @VisibleForTesting
    public static final Integer MAX_PAGES = 23_000;
    @Max(value = MAX_PAGES)
    private Integer numberOfPages;
    @Max(value = MAX_PAGES)
    private Integer pagesRead;`
In this case as well I get same java error :(

Comment (amit1307) [reply src/main/java/com/karankumar/bookproject/backend/entity/Book.java:52]: Hey how about something like below:

`public static final int MAX_PAGES = 23000
    @NotNull
    @NotEmpty
    private String title;
    @Max(value = MAX_PAGES)
    private Integer numberOfPages;
    @Max(value = MAX_PAGES)`

File: src/test/java/com/karankumar/bookproject/backend/service/BookServiceTest.java
Comment (amit1307) [reply src/test/java/com/karankumar/bookproject/backend/service/BookServiceTest.java:101]: Ah okay, I got it now..cool, that's fine then.

File: src/main/java/com/karankumar/bookproject/backend/entity/Book.java
Comment (prabhjot018) [reply src/main/java/com/karankumar/bookproject/backend/entity/Book.java]: Annotation allows primitive values, changing Integer to int helped.

**Review by knjk04 — APPROVED** [review]
All good, thanks for this!

**Review by amit1307 — APPROVED** [review]
lgtm

**Pull request merged**
